src/couch_stats/src/csrt_server.erl

```diff
@@ -0,0 +1,198 @@
+new_context(Type, Nonce) ->
+    #rctx{
+       nonce = Nonce,
+       pid_ref = create_pid_ref(),
+       type = Type
+    }.
+
+-spec set_context_dbname(DbName, PidRef) -> boolean() when
+    DbName :: dbname(), PidRef :: maybe_pid_ref().
+set_context_dbname(_, undefined) ->
+    false;
+set_context_dbname(DbName, PidRef) ->
+    update_element(PidRef, [{#rctx.dbname, DbName}]).
+
+%%set_context_handler_fun(_, undefined) ->
+%%    ok;
+%%set_context_handler_fun(Fun, PidRef) when is_function(Fun) ->
+%%    FProps = erlang:fun_info(Fun),
+%%    Mod = proplists:get_value(module, FProps),
+%%    Func = proplists:get_value(name, FProps),
+%%    #rctx{type=#coordinator{}=Coordinator} = get_resource(PidRef),
+%%    Update = [{#rctx.type, Coordinator#coordinator{mod=Mod, func=Func}}],
+%%    update_element(PidRef, Update).
+
+-spec set_context_username(UserName, PidRef) -> boolean() when
+    UserName :: username(), PidRef :: maybe_pid_ref().
+set_context_username(_, undefined) ->
+    ok;
+set_context_username(UserName, PidRef) ->
+    update_element(PidRef, [{#rctx.username, UserName}]).
+
+-spec get_context_type(Rctx :: rctx()) -> rctx_type().
+get_context_type(#rctx{type=Type}) ->
+    Type.
+
+-spec set_context_type(Type, PidRef) -> boolean() when
+    Type :: rctx_type(), PidRef :: maybe_pid_ref().
+set_context_type(Type, PidRef) ->
+    update_element(PidRef, [{#rctx.type, Type}]).
+
+-spec create_resource(Rctx :: rctx()) -> true.
+create_resource(#rctx{} = Rctx) ->
+    catch ets:insert(?MODULE, Rctx).
+
+-spec destroy_resource(PidRef :: maybe_pid_ref()) -> boolean().
+destroy_resource(undefined) ->
+    false;
+destroy_resource({_,_}=PidRef) ->
+    catch ets:delete(?MODULE, PidRef).
+
+-spec get_resource(PidRef :: maybe_pid_ref()) -> maybe_rctx().
+get_resource(undefined) ->
+    undefined;
+get_resource(PidRef) ->
+    catch case ets:lookup(?MODULE, PidRef) of
+        [#rctx{}=Rctx] ->
+            Rctx;
+        [] ->
+            undefined
+    end.
+
+-spec is_rctx_field(Field :: rctx_field() | atom()) -> boolean().
+is_rctx_field(Field) ->
+    maps:is_key(Field, ?KEYS_TO_FIELDS).
+
+-spec get_rctx_field(Field :: rctx_field()) -> non_neg_integer().
+get_rctx_field(Field) ->
+    maps:get(Field, ?KEYS_TO_FIELDS).
+
+-spec update_counter(PidRef, Field, Count) -> non_neg_integer() when
+        PidRef :: maybe_pid_ref(),
+        Field :: rctx_field(),
+        Count :: non_neg_integer().
+update_counter(undefined, _Field, _Count) ->
+    0;
+update_counter({_Pid,_Ref}=PidRef, Field, Count) when Count >= 0 ->
+    %% TODO: mem3 crashes without catch, why do we lose the stats table?
+    case is_rctx_field(Field) of
+        true ->
+            Update = {get_rctx_field(Field), Count},
+            catch ets:update_counter(?MODULE, PidRef, Update, #rctx{pid_ref=PidRef});
+        false ->
+            0
+    end.
+
+-spec inc(PidRef :: maybe_pid_ref(), Field :: rctx_field()) -> non_neg_integer().
```

Comment:
consider changing result type to be `ok`.

Reply:
This intentionally returns the value from `ets:update_counter` as ets returns the atomically updated new value so by bubbling it up we get to read it for free. More details here: https://github.com/apache/couchdb/pull/5491#discussion_r2059397392